diff -ur gnutls-3.3.8-orig/m4/hooks.m4 gnutls-3.3.8/m4/hooks.m4
--- gnutls-3.3.8-orig/m4/hooks.m4	2014-09-18 07:54:45.000000000 +0200
+++ gnutls-3.3.8/m4/hooks.m4	2014-09-18 19:08:19.223172863 +0200
@@ -104,7 +104,7 @@
       included_libtasn1=$withval,
       included_libtasn1=no)
   if test "$included_libtasn1" = "no"; then
-    PKG_CHECK_MODULES(LIBTASN1, [libtasn1 >= 3.9], [], [included_libtasn1=yes])
+    PKG_CHECK_MODULES(LIBTASN1, [libtasn1 >= 3.8], [], [included_libtasn1=yes])
     if test "$included_libtasn1" = yes; then
       AC_MSG_WARN([[
   *** 
Only in gnutls-3.3.8/m4: hooks.m4~
