diff --git a/lib/gnutls_sig.c b/lib/gnutls_sig.c
index c32d9c3..6a11701 100644
--- a/lib/gnutls_sig.c
+++ b/lib/gnutls_sig.c
@@ -244,8 +244,8 @@ sign_tls_hash (gnutls_session_t session, gnutls_digest_algorithm_t hash_algo,
               ret = _gnutls_set_datum(&digest, hash_concat->data, hash_concat->size);
               if (ret < 0)
                 return gnutls_assert_val(ret);
-              
-              ret = pk_prepare_hash (gnutls_privkey_get_pk_algorithm(pkey, NULL), hash_algo, &digest);
+
+              ret = pk_prepare_hash (cert->subject_pk_algorithm, hash_algo, &digest);
               if (ret < 0)
                 {
                   gnutls_assert ();
