diff -ur gnutls-2.8.5/lib/includes/gnutls/x509.h gnutls-2.8.5.new/lib/includes/gnutls/x509.h
--- gnutls-2.8.5/lib/includes/gnutls/x509.h	2009-06-02 21:07:14.000000000 +0200
+++ gnutls-2.8.5.new/lib/includes/gnutls/x509.h	2015-01-16 14:38:56.168991623 +0100
@@ -511,7 +511,12 @@
      * periods of certificate chains. Don't set this unless you
      * understand the security implications.
      */
-    GNUTLS_VERIFY_DISABLE_TIME_CHECKS = 64
+    GNUTLS_VERIFY_DISABLE_TIME_CHECKS = 64,
+
+    /* If set a signer in the trusted list is never checked for expiration or
+     * activation.
+     */
+    GNUTLS_VERIFY_DISABLE_TRUSTED_TIME_CHECKS = 128
   } gnutls_certificate_verify_flags;
 
   int gnutls_x509_crt_check_issuer (gnutls_x509_crt_t cert,
diff -ur gnutls-2.8.5/lib/x509/verify.c gnutls-2.8.5.new/lib/x509/verify.c
--- gnutls-2.8.5/lib/x509/verify.c	2015-01-16 14:42:05.206530174 +0100
+++ gnutls-2.8.5.new/lib/x509/verify.c	2015-01-16 14:40:23.551702820 +0100
@@ -41,7 +41,8 @@
 static int _gnutls_verify_certificate2 (gnutls_x509_crt_t cert,
 					const gnutls_x509_crt_t * trusted_cas,
 					int tcas_size, unsigned int flags,
-					unsigned int *output);
+					unsigned int *output,
+                                        gnutls_x509_crt_t *issuer);
 
 static int is_crl_issuer (gnutls_x509_crl_t crl,
 			  gnutls_x509_crt_t issuer_cert);
@@ -259,17 +260,17 @@
  * 'flags': an OR of the gnutls_certificate_verify_flags enumeration.
  *
  * Output will hold some extra information about the verification
- * procedure.
+ * procedure. Issuer will hold the actual issuer from the trusted list.
  */
 static int
 _gnutls_verify_certificate2 (gnutls_x509_crt_t cert,
 			     const gnutls_x509_crt_t * trusted_cas,
 			     int tcas_size, unsigned int flags,
-			     unsigned int *output)
+			     unsigned int *output, gnutls_x509_crt_t *_issuer)
 {
   gnutls_datum_t cert_signed_data = { NULL, 0 };
   gnutls_datum_t cert_signature = { NULL, 0 };
-  gnutls_x509_crt_t issuer;
+  gnutls_x509_crt_t issuer = NULL;
   int ret, issuer_version, result = 0;
 
   if (output)
@@ -296,6 +297,8 @@
       return 0;
     }
 
+  if (_issuer != NULL) *_issuer = issuer;
+
   issuer_version = gnutls_x509_crt_get_version (issuer);
   if (issuer_version < 0)
     {
@@ -398,6 +401,29 @@
   return is_issuer (cert, issuer);
 }
 
+static unsigned int check_time(gnutls_x509_crt_t crt, time_t now)
+{
+  int status = 0;
+  time_t t;
+
+  t = gnutls_x509_crt_get_activation_time (crt);
+  if (t == (time_t) - 1 || now < t)
+    {
+      status |= GNUTLS_CERT_NOT_ACTIVATED;
+      status |= GNUTLS_CERT_INVALID;
+      return status;
+    }
+
+  t = gnutls_x509_crt_get_expiration_time (crt);
+  if (t == (time_t) - 1 || now > t)
+    {
+      status |= GNUTLS_CERT_EXPIRED;
+      status |= GNUTLS_CERT_INVALID;
+      return status;
+    }
+
+  return 0;
+}
 
 /* Verify X.509 certificate chain.
  *
@@ -416,6 +442,8 @@
 {
   int i = 0, ret;
   unsigned int status = 0, output;
+  time_t now = time (0);
+  gnutls_x509_crt_t issuer = NULL;
 
   if (clist_size > 1)
     {
@@ -451,6 +479,16 @@
 	      if (check_if_same_cert (certificate_list[i],
 				      trusted_cas[j]) == 0)
 		{
+		  /* explicity time check for trusted CA that we remove from
+		   * list. GNUTLS_VERIFY_DISABLE_TRUSTED_TIME_CHECKS
+                   */
+		  if (!(flags&GNUTLS_VERIFY_DISABLE_TRUSTED_TIME_CHECKS) &&
+                      !(flags&GNUTLS_VERIFY_DISABLE_TIME_CHECKS))
+                    {
+                      status |= check_time(trusted_cas[j], now);
+                      if (status != 0)
+                        return status;
+                    }
 		  clist_size = i;
 		  break;
 		}
@@ -466,7 +504,8 @@
    * in self signed etc certificates.
    */
   ret = _gnutls_verify_certificate2 (certificate_list[clist_size - 1],
-				     trusted_cas, tcas_size, flags, &output);
+				     trusted_cas, tcas_size, flags, &output,
+                                     &issuer);
   if (ret == 0)
     {
       /* if the last certificate in the certificate
@@ -499,23 +538,27 @@
    */
   if (!(flags & GNUTLS_VERIFY_DISABLE_TIME_CHECKS))
     {
-      time_t t, now = time (0);
+      /* check the time of the issuer first */
+      if (!(flags&GNUTLS_VERIFY_DISABLE_TRUSTED_TIME_CHECKS))
+        {
+          if (issuer == NULL)
+            {
+              gnutls_assert();
+              return (GNUTLS_CERT_INVALID);
+            }
+
+          status |= check_time(issuer, now);
+          if (status != 0)
+            {
+              return status;
+            }
+        }
 
       for (i = 0; i < clist_size; i++)
 	{
-	  t = gnutls_x509_crt_get_activation_time (certificate_list[i]);
-	  if (t == (time_t) -1 || now < t)
-	    {
-	      status |= GNUTLS_CERT_NOT_ACTIVATED;
-	      status |= GNUTLS_CERT_INVALID;
-	      return status;
-	    }
-
-	  t = gnutls_x509_crt_get_expiration_time (certificate_list[i]);
-	  if (t == (time_t) -1 || now > t)
+          status |= check_time(certificate_list[i], now);
+          if (status != 0)
 	    {
-	      status |= GNUTLS_CERT_EXPIRED;
-	      status |= GNUTLS_CERT_INVALID;
 	      return status;
 	    }
 	}
@@ -536,7 +579,7 @@
       if ((ret =
 	   _gnutls_verify_certificate2 (certificate_list[i - 1],
 					&certificate_list[i], 1, flags,
-					NULL)) == 0)
+					NULL, NULL)) == 0)
 	{
 	  status |= GNUTLS_CERT_INVALID;
 	  return status;
@@ -994,18 +1037,12 @@
 			int CA_list_length, unsigned int flags,
 			unsigned int *verify)
 {
-  int ret;
   /* Verify certificate 
    */
-  ret =
-    _gnutls_verify_certificate2 (cert, CA_list, CA_list_length, flags,
-				 verify);
-  if (ret < 0)
-    {
-      gnutls_assert ();
-      return ret;
-    }
-
+  *verify =
+    _gnutls_x509_verify_certificate (&cert, 1,
+				     CA_list, CA_list_length, NULL,
+				     0, flags);
   return 0;
 }
 
