diff --git a/lib/gnutls_priority.c b/lib/gnutls_priority.c
index 622ef55..1e0e1c1 100644
--- a/lib/gnutls_priority.c
+++ b/lib/gnutls_priority.c
@@ -253,7 +253,6 @@ static const int cipher_priority_performance[] = {
   GNUTLS_CIPHER_CAMELLIA_128_CBC,
 #endif
   GNUTLS_CIPHER_AES_128_CBC,
-  GNUTLS_CIPHER_3DES_CBC,
   GNUTLS_CIPHER_AES_256_CBC,
 #ifdef	ENABLE_CAMELLIA
   GNUTLS_CIPHER_CAMELLIA_256_CBC,
