diff --git a/src/serv.c b/src/serv.c
index 06dedcd..c2ef41b 100644
--- a/src/serv.c
+++ b/src/serv.c
@@ -50,6 +50,8 @@
 #include "minmax.h"
 #include "sockets.h"
 
+#define _GNUTLS_E_UNRECOGNIZED_NAME -294
+
 /* konqueror cannot handle sending the page in multiple
  * pieces.
  */
@@ -388,7 +390,7 @@ post_client_hello(gnutls_session_t session)
 
 	if (sni_hostname_fatal == 1) {
 		/* abort the connection, propagate error up the stack */
-		ret = -1;
+		ret = _GNUTLS_E_UNRECOGNIZED_NAME;
 		goto end;
 	}
 
@@ -926,6 +928,15 @@ tls_log_func (int level, const char *str)
 
 static void gaa_parser (int argc, char **argv);
 
+int _gnutls_alert_send_appropriate (gnutls_session_t session, int err)
+{
+  if (err == _GNUTLS_E_UNRECOGNIZED_NAME)
+    return gnutls_alert_send(session,
+                             GNUTLS_AL_FATAL,
+                             GNUTLS_A_UNRECOGNIZED_NAME);
+  return gnutls_alert_send_appropriate(session, err);
+}
+
 int
 main (int argc, char **argv)
 {
@@ -1288,7 +1299,7 @@ main (int argc, char **argv)
                     do
                       {
                         ret =
-                          gnutls_alert_send_appropriate (j->tls_session, r);
+                          _gnutls_alert_send_appropriate (j->tls_session, r);
                       }
                     while (ret == GNUTLS_E_AGAIN
                            || ret == GNUTLS_E_INTERRUPTED);
@@ -1337,7 +1348,7 @@ main (int argc, char **argv)
                           {
                             do
                               {
-                                ret = gnutls_alert_send_appropriate
+                                ret = _gnutls_alert_send_appropriate
                                   (j->tls_session, r);
                               }
                             while (ret == GNUTLS_E_AGAIN
@@ -1357,7 +1368,7 @@ main (int argc, char **argv)
                             fprintf (stderr, "Error while receiving data\n");
                             do
                               {
-                                ret = gnutls_alert_send_appropriate(j->tls_session, r);
+                                ret = _gnutls_alert_send_appropriate(j->tls_session, r);
                               }
                             while(ret == GNUTLS_E_AGAIN || ret == GNUTLS_E_INTERRUPTED);
                             GERR (r);
@@ -1422,7 +1433,7 @@ main (int argc, char **argv)
                     do
                       {
                         ret =
-                          gnutls_alert_send_appropriate (j->tls_session, r);
+                          _gnutls_alert_send_appropriate (j->tls_session, r);
                       }
                     while (ret == GNUTLS_E_AGAIN);
                   }
