diff --git a/lib/accelerated/x86/sha-x86-ssse3.c b/lib/accelerated/x86/sha-x86-ssse3.c
index 17724f7..e9d8eab 100644
--- a/lib/accelerated/x86/sha-x86-ssse3.c
+++ b/lib/accelerated/x86/sha-x86-ssse3.c
@@ -248,7 +248,7 @@ static int _ctx_init(gnutls_digest_algorithm_t algo,
 	case GNUTLS_DIG_SHA224:
 		sha224_init(&ctx->ctx.sha224);
 		ctx->update = (update_func) x86_sha256_update;
-		ctx->digest = (digest_func) sha256_digest;
+		ctx->digest = (digest_func) sha224_digest;
 		ctx->init = (init_func) sha224_init;
 		ctx->ctx_ptr = &ctx->ctx.sha224;
 		ctx->length = SHA224_DIGEST_SIZE;
