diff --git a/lib/ext_signature.c b/lib/ext_signature.c
index 3e81fa0..46cf287 100644
--- a/lib/ext_signature.c
+++ b/lib/ext_signature.c
@@ -167,9 +167,9 @@ _gnutls_sign_algorithm_parse_data (gnutls_session_t session,
                 continue;
             }
 
-          priv->sign_algorithms[priv->sign_algorithms_size++] = sig;
           if (priv->sign_algorithms_size == MAX_SIGNATURE_ALGORITHMS)
             break;
+          priv->sign_algorithms[priv->sign_algorithms_size++] = sig;
         }
     }
 
@@ -211,7 +211,7 @@ _gnutls_signature_algorithm_recv_params (gnutls_session_t session,
     {
       /* SERVER SIDE - we must check if the sent cert type is the right one
        */
-      if (data_size > 2)
+      if (data_size >= 2)
         {
           uint16_t len;
 
@@ -306,8 +306,7 @@ _gnutls_session_get_sign_algo (gnutls_session_t session, gnutls_cert* cert)
                                   &epriv);
   priv = epriv.ptr;
 
-  if (ret < 0 || !_gnutls_version_has_selectable_sighash (ver)
-      || priv->sign_algorithms_size == 0)
+  if (ret < 0 || !_gnutls_version_has_selectable_sighash (ver))
     /* none set, allow SHA-1 only */
     {
       return _gnutls_x509_pk_to_sign (cert->subject_pk_algorithm, GNUTLS_DIG_SHA1);
diff --git a/lib/gnutls_alert.c b/lib/gnutls_alert.c
index affdff2..298d57d 100644
--- a/lib/gnutls_alert.c
+++ b/lib/gnutls_alert.c
@@ -210,6 +210,7 @@ gnutls_error_to_alert (int err, int *level)
     case GNUTLS_E_UNSUPPORTED_SIGNATURE_ALGORITHM:
     case GNUTLS_E_SAFE_RENEGOTIATION_FAILED:
     case GNUTLS_E_INCOMPAT_DSA_KEY_WITH_TLS_PROTOCOL:
+    case GNUTLS_E_UNKNOWN_PK_ALGORITHM:
       ret = GNUTLS_A_HANDSHAKE_FAILURE;
       _level = GNUTLS_AL_FATAL;
       break;
