diff --git a/lib/gnutls_algorithms.c b/lib/gnutls_algorithms.c
index f29f99e..e3eaa66 100644
--- a/lib/gnutls_algorithms.c
+++ b/lib/gnutls_algorithms.c
@@ -1936,6 +1936,8 @@ static const gnutls_sign_entry sign_algorithms[] = {
   {"GOST R 34.10-2001", SIG_GOST_R3410_2001_OID, 0, 0, 0,
    TLS_SIGN_AID_UNKNOWN},
   {"GOST R 34.10-94", SIG_GOST_R3410_94_OID, 0, 0, 0, TLS_SIGN_AID_UNKNOWN},
+  {"RSA-SHA", SIG_RSA_SHA1_OID, GNUTLS_SIGN_RSA_SHA1, GNUTLS_PK_RSA,
+   GNUTLS_MAC_SHA1, {2, 1}}, /* compat */
   {0, 0, 0, 0, 0, TLS_SIGN_AID_UNKNOWN}
 };
 
