diff -up gnutls-2.8.5/src/Makefile.am.aliasing gnutls-2.8.5/src/Makefile.am
--- gnutls-2.8.5/src/Makefile.am.aliasing	2009-06-02 20:59:32.000000000 +0200
+++ gnutls-2.8.5/src/Makefile.am	2012-10-10 14:32:51.842330717 +0200
@@ -19,7 +19,7 @@
 
 SUBDIRS = cfg
 
-AM_CFLAGS = $(WARN_CFLAGS) $(WERROR_CFLAGS)
+AM_CFLAGS = $(WARN_CFLAGS) $(WERROR_CFLAGS) -fno-strict-aliasing
 AM_CPPFLAGS = \
 	-I$(srcdir)/../gl			\
 	-I$(builddir)/../gl			\
diff -up gnutls-2.8.5/src/Makefile.in.aliasing gnutls-2.8.5/src/Makefile.in
--- gnutls-2.8.5/src/Makefile.in.aliasing	2009-11-02 11:36:01.000000000 +0100
+++ gnutls-2.8.5/src/Makefile.in	2012-10-10 14:33:11.019741961 +0200
@@ -813,7 +813,7 @@ top_build_prefix = @top_build_prefix@
 top_builddir = @top_builddir@
 top_srcdir = @top_srcdir@
 SUBDIRS = cfg
-AM_CFLAGS = $(WARN_CFLAGS) $(WERROR_CFLAGS)
+AM_CFLAGS = $(WARN_CFLAGS) $(WERROR_CFLAGS) -fno-strict-aliasing
 AM_CPPFLAGS = \
 	-I$(srcdir)/../gl			\
 	-I$(builddir)/../gl			\
