diff --git a/lib/gnutls_priority.c b/lib/gnutls_priority.c
index bd8cb5a..f4689e6 100644
--- a/lib/gnutls_priority.c
+++ b/lib/gnutls_priority.c
@@ -543,7 +543,7 @@ gnutls_priority_init (gnutls_priority_t * priority_cache,
    * when we make it the default.
    */
   (*priority_cache)->sr = SR_PARTIAL;
-  (*priority_cache)->ssl3_record_version = 1;
+  (*priority_cache)->ssl3_record_version = 0;
 
   if (priorities == NULL)
     priorities = "NORMAL";
